bundles/org.eclipse.emfcloud.modelserver.common/src/org/eclipse/emfcloud/modelserver/common/patch/AbstractJsonPatchHelper.java

```diff
@@ -187,31 +187,35 @@ protected Command getRemoveCommand(final String modelURI, final ResourceSet reso
       String jsonPath = path.asText();
       SettingValue setting = getSetting(modelURI, resourceSet, jsonPath);
 
+      // If neither the index nor the feature are present, we delete an Object
+      EObject eObject = setting.getEObject();
+
       // If the index is present, we remove a value from a list
       if (setting.getIndex().isPresent()) {
          int index = setting.getIndex().get();
          if (index == CommandParameter.NO_INDEX) {
             // If the index is "-", remove the last value from the list
-            Object currentValue = setting.getEObject().eGet(setting.getFeature());
+            Object currentValue = eObject.eGet(setting.getFeature());
             if (currentValue instanceof Collection) {
                index = ((Collection<?>) currentValue).size() - 1;
             }
          }
-         return RemoveCommand.create(getEditingDomain(setting.getEObject()), setting.getEObject(),
-            setting.getFeature(), index);
+         return RemoveCommand.create(getEditingDomain(eObject), eObject, setting.getFeature(), index);
+      }
+
+      // if we are in a feature which is a list, remove the element from said feature
+      if (eObject.eContainingFeature() != null && eObject.eContainingFeature().isMany()) {
```

Comment:
Hmm, looking at the doc: https://download.eclipse.org/modeling/emf/emf/javadoc/2.9.0/org/eclipse/emf/ecore/EObject.html#eContainingFeature()
I would claim it doesn't matter and especially with the check for `isMany` it should be fine. The docu says for 
- eContainingFeature: `Returns the particular feature of the container that actually holds the object ... this feature may be an attribute representing a feature map ...`
- eContainmentFeature: `Returns the containment feature that properly contains the object ... this feature may not be a direct feature of the container's class ...`

Or do I misunderstand the docu?

Reply:
It will work for objects that are not contained by feature maps. But it is an API that was added for the implementation of feature maps. In the case that the object being deleted is contained in a feature map, the `eContainingFeature` will be the feature-map `eAttribute` and it will often have upper bound -2 (`ETypedElement.UNSPECIFIED_MULTIPLICITY`). And in this case you would actually have to use the `FeatureMapUtil::isMany(...)` API to get the true "is many" answer. And I'm not sure how any of that would work with the commands that we're trying to create for patches.

For maximum clarity, I would prefer to use the `eContainmentFeature()` API that we use everywhere else, which is typed as an `EReference`, so as not to confuse readers/maintainers that the JSON Patch support handles feature maps, because it doesn't try to.